core/inspector.rs

```diff
@@ -259,94 +314,82 @@ impl JsRuntimeInspector {
             w.poll_state = PollState::Idle;
-            // Register the task waker that can be used to wake the parent
-            // task that will poll the inspector future.
-            if let Some(cx) = invoker_cx.take() {
-              w.task_waker.replace(cx.waker().clone());
-            }
-            // Register the address of the inspector, which allows the waker
-            // to request an interrupt from the isolate.
-            w.inspector_ptr = NonNull::new(self as *const _ as *mut Self);
-          }
-          PollState::Polling if should_block => {
-            // Isolate execution has been paused but there are no more
-            // events to process, so this thread will be parked. Therefore,
-            // store the current thread handle in the waker so it knows
-            // which thread to unpark when new events arrive.
-            w.poll_state = PollState::Parked;
-            w.parked_thread.replace(thread::current());
+            // Capture the waker that, when used, will get the inspector polled.
+            w.task_waker.replace(invoker_cx.waker().clone());
           }
           _ => unreachable!(),
         };
         w.poll_state
       });
-      match new_state {
-        PollState::Idle => break Ok(Poll::Pending), // Yield to task.
-        PollState::Polling => {} // Poll the session handler again.
-        PollState::Parked => thread::park(), // Park the thread.
+
+      match new_poll_state {
+        PollState::Idle => break Poll::Pending,
+        PollState::Polling => continue, // Poll the session handler again.
         _ => unreachable!(),
       };
     }
   }
 
+  /// Accepts incoming connections from inspector clients, and polls established
+  /// inspector sessions for messages that need to be dispatched to V8. This
+  /// function will repeatedly poll its innner streams and will not return until
+  /// they all yield `Pending` or have ended.
+  fn poll_sessions_inner(&self, cx: &mut Context) {
+    loop {
+      let mut sessions = self.sessions.borrow_mut();
+
+      // Accept new connections.
+      let poll_result = sessions.session_rx.poll_next_unpin(cx);
+      match poll_result {
+        Poll::Ready(Some(session_proxy)) => {
+          let session = InspectorSession::new(
+            self.v8_inspector.clone(),
+            session_proxy,
+            false,
+          );
+          sessions.established.push(session);
+          // `session_rx` needs to be polled repeatedly until it is `Pending`.
+          continue;
+        }
+        Poll::Ready(None) => unreachable!(), // `session_rx` should never end.
+        Poll::Pending => {}
+      }
+
+      // Poll established inspector sessions.
+      let poll_result = sessions.established.poll_next_unpin(cx);
+      if let Poll::Ready(Some(session_stream_item)) = poll_result {
+        let (v8_session_ptr, msg) = session_stream_item;
+        // Don't hold the borrow on `sessions` while dispatching a message.
```

Comment:
Note to self: update comment to `// Don't hold the borrow on `sessions` while dispatching a message as it might result in a call to `poll_sessions_sync`.`